File: src/main/java/org/owari/shigure/Expression.java
package org.owari.shigure;

import org.owari.shigure.impl.*;
import org.owari.shigure.util.SneakyThrow;

import java.util.function.Function;

public final class Expression implements Function<Context, Double> {
    public static Expression of(String source) {
        return new Expression(source);
    }
    private static final int JITThreshold = 16_000;

    private final String source;
    private int count = 0;
    private boolean isInit = false;
    private boolean isJIT = false;
    private OpcodeSet os;
    private CalcFunc impl;

    private Expression(String source) {
        this.source = source;
    }

    private void init() {
        Tokenizer t = new Tokenizer(source);
        Parser p = new Parser(t);
        os = p.parse();
    }

    public double invoke(Context ctx) {
        if(isJIT) return impl.invoke(ctx);
        if(!isInit) {
            init();
            isInit = true;
        }
        count++;
        if(count >= JITThreshold) {
            compile();
            return impl.invoke(ctx);
        }
        return Interpreter.eval(os, ctx);
    }

    public void compile() {
        if(isJIT) return;
        if(!isInit) {
            init();
            isInit = true;
        }
        Assembler as = Shigure.getDefaultAssembler();
        String name = as.newClassName();
        CodeGenerator cg = new CodeGenerator(os, name);
        try {
            impl = (CalcFunc) as.assemble(name, cg.generate()).newInstance();
        } catch (Throwable t) {
            throw SneakyThrow.sneakyThrow(t);
        }
        isJIT = true;
    }

    @Override
    public Double apply(Context ctx) {
        return invoke(ctx);
    }
}

File: src/main/kotlin/org/owari/shigure/Expression.kt
package org.owari.shigure

import org.owari.shigure.impl.*

@FunctionalInterface
interface Expression : (Context) -> Double {
    override operator fun invoke(ctx: Context): Double

    companion object {
        inline fun of(s: String) = ExpressionImpl(s)
    }
}

interface JITExpression : Expression {
    fun compile()
}

class ExpressionImpl(private val source: String) : JITExpression {
    companion object {
        // benchmarked
        const val JITThreshold = 16_000
    }

    private var count = 0
    private var isJIT = false

    private val tokenizer = PassiveTokenizer(source)
    private val parser = Parser(tokenizer)
    private val ast by lazy { parser.parse() }
    private val astEvaluator: ASTEvaluator by lazy { ASTEvaluator(ast) }
    private val impl: CalculateFunc by lazy { TODO() }

    override operator fun invoke(ctx: Context): Double  {
        if(isJIT) return impl.eval(ctx)
        count++
        if(count > JITThreshold) {
            isJIT = true
            return impl.eval(ctx)
        }
        return astEvaluator.eval(ctx)
    }

    override fun compile() {
        this.impl
        isJIT = true
    }
}

/**
 * 用于计算的函数
 * 为了 asm 生成代码方便, 设置了插桩父类
 */
abstract class CalculateFunc {
    abstract fun eval(ctx: Context): Double
}
